File: Python/number_complement.py
# https://leetcode.com/problems/number-complement/description/
from utils.examination import Examiner, Testcase, Logger


class Solution:
    def findComplement(self, num: int) -> int:
        b: int = 0
        i: int = 0
        p: int = 0
        while num >= p:
            b += p
            p = 2 ** i
            i += 1
        return num ^ b


def examine() -> None:
    Examiner(Solution(), 'findComplement').run((
        Testcase((5,), 2),
        Testcase((1,), 0),
    ))

File: TypeScript/solutions/number-complement.ts
// https://leetcode.com/problems/number-complement/description/
import { Examiner, Testcase } from '../utils';


function findComplement(num: number): number {
    let b = 0, i = 0, p = 0;
    while (num >= p) { b += p; p = 2 ** i; i++; }
    return num ^ b;
}

export function examine(): void {
    new Examiner(findComplement).run([
        new Testcase([5], 2),
        new Testcase([1], 0),
    ]);
}
